File: androidApp/src/main/kotlin/com/snapp/android/ui/widget/WidgetRegistry.kt
package com.snapp.android.ui.widget

import androidx.compose.runtime.Composable
import com.snapp.android.ui.widget.renderer.BarChartWidget
import com.snapp.android.ui.widget.renderer.CardWidget
import com.snapp.android.ui.widget.renderer.FieldWidget
import com.snapp.android.ui.widget.renderer.ImageWidget
import com.snapp.android.ui.widget.renderer.LineChartWidget
import com.snapp.android.ui.widget.renderer.MetricWidget
import com.snapp.android.ui.widget.renderer.PieChartWidget
import com.snapp.android.ui.widget.renderer.TableWidget

/** Registry: type String → @Composable renderer. See ARCHITECTURE ui/widget/. */
object WidgetRegistry {
    private val builders = mutableMapOf<String, @Composable (WidgetRenderContext) -> Unit>()

    fun register(type: String, builder: @Composable (WidgetRenderContext) -> Unit) {
        builders[type] = builder
    }

    fun get(type: String): (@Composable (WidgetRenderContext) -> Unit)? = builders[type]
}

fun registerWidgets() {
    val registry = WidgetRegistry // import

    // Card
    registry.register("card") { CardWidget(it) }
    registry.register("record") { CardWidget(it) }

    // Simple widgets
    registry.register("field") { FieldWidget(it) }
    registry.register("image") { ImageWidget(it) }
    registry.register("metric") { MetricWidget(it) }

    // Charts
    registry.register("bar") { BarChartWidget(it) }
    registry.register("bar-chart") { BarChartWidget(it) }
    registry.register("pie") { PieChartWidget(it) }
    registry.register("pie-chart") { PieChartWidget(it) }
    registry.register("donut") { PieChartWidget(it) }
    registry.register("donut-chart") { PieChartWidget(it) }
    registry.register("line") { LineChartWidget(it) }
    registry.register("line-chart") { LineChartWidget(it) }
    registry.register("area") { LineChartWidget(it) }
    registry.register("area-chart") { LineChartWidget(it) }

    // Table/List
    registry.register("table") { TableWidget(it) }
    registry.register("list") { TableWidget(it) }
}

File: iosApp/iosApp/Widget/WidgetRegistry.swift
import SwiftUI
import shared

/// Widget registry — same role as web client/src/lib/widgetRegistry.ts (Map<type, Component>).
/// WidgetRenderer gets component via widgetRegistry.get(widget.type); we get builder via get(type) and call with WidgetRenderContext.
/// Keys must match API widget.type and web registry exactly (bar-chart, card, tabs, field, etc.). Web has duplicate keys (stepper, report); we register each once.
final class WidgetRegistry {
    static let shared = WidgetRegistry()
    private var builders: [String: (WidgetRenderContext) -> AnyView] = [:]

    private init() {
        registerAllWidgets()
    }

    func register(_ type: String, builder: @escaping (WidgetRenderContext) -> AnyView) {
        builders[type] = builder
    }

    func get(_ type: String) -> ((WidgetRenderContext) -> AnyView)? {
        builders[type]
    }

    func registerAllWidgets() {
        // Charts
        register("bar-chart") { AnyView(BarChartWidgetView(context: $0)) }
        register("line-chart") { AnyView(LineChartWidgetView(context: $0)) }
        register("area-chart") { AnyView(AreaChartWidgetView(context: $0)) }
        register("pie-chart") { AnyView(PieChartWidgetView(context: $0)) }
        register("donut-chart") { AnyView(DonutChartWidgetView(context: $0)) }
        register("radar-chart") { AnyView(RadarChartWidgetView(context: $0)) }
        register("radial-chart") { AnyView(RadialChartWidgetView(context: $0)) }
        register("bubble-chart") { AnyView(BubbleChartWidgetView(context: $0)) }
        register("funnel-chart") { AnyView(FunnelChartWidgetView(context: $0)) }
        register("candle-chart") { AnyView(CandlestickWidgetView(context: $0)) }
        register("heatmap") { AnyView(HeatMapWidgetView(context: $0)) }

        // Data / list / table
        register("list") { AnyView(TableWidgetView(context: $0)) }
        register("table") { AnyView(ComprehensiveTableWidgetView(context: $0)) }

        // Containers
        register("card") { AnyView(CardWidgetView(context: $0)) }
        register("record") { AnyView(CardWidgetView(context: $0)) }
        register("accordion") { AnyView(AccordionWidgetView(context: $0)) }
        register("tabs") { AnyView(TabWidgetView(context: $0)) }

        // KPI / metric
        register("metric") { AnyView(MetricWidgetView(context: $0)) }

        // Other widgets (exact keys as web)
        register("calendar") { AnyView(CalendarWidgetView(context: $0)) }
        register("chat") { AnyView(ChatWidgetView(context: $0)) }
        register("todo") { AnyView(TodoListWidgetView(context: $0)) }
        register("field") { AnyView(FieldWidgetView(context: $0)) }
        register("importer") { AnyView(ImportWidgetView(context: $0)) }
        register("contact") { AnyView(ContactWidgetView(context: $0)) }
        register("report") { AnyView(ReportsWidgetView(context: $0)) }
        register("rich-text-editor") { AnyView(RichTextEditorWidgetView(context: $0)) }
        register("google-map") { AnyView(GoogleMapWidgetView(context: $0)) }
        register("whiteboard") { AnyView(WhiteboardWidgetView(context: $0)) }
        register("kanban") { AnyView(KanbanWidgetView(context: $0)) }
        register("gantt-chart") { AnyView(GanttChartWidgetView(context: $0)) }
        register("stepper") { AnyView(StepperWidgetView(context: $0)) }
        register("payment-gateway") { AnyView(PaymentGatewayWidgetView(context: $0)) }
        register("carousel") { AnyView(CarouselWidgetView(context: $0)) }
        register("file-upload") { AnyView(FileUploadWidgetView(context: $0)) }
        register("template-builder") { AnyView(TemplateBuilderWidgetView(context: $0)) }
        register("rich-text-display") { AnyView(RichTextDisplayWidgetView(context: $0)) }
        register("image") { AnyView(ImageDisplayWidgetView(context: $0)) }
        register("video") { AnyView(VideoDisplayWidgetView(context: $0)) }
        register("button") { AnyView(ButtonWidgetView(context: $0)) }
        register("file-viewer") { AnyView(FileViewerWidgetView(context: $0)) }
        register("pipeline") { AnyView(PipelineWidgetView(context: $0)) }
        register("activity-feed") { AnyView(ActivityFeedWidgetView(context: $0)) }
        register("i-frame") { AnyView(IFrameWidgetView(context: $0)) }
        register("knowledge-graph") { AnyView(KnowledgeGraphWidgetView(context: $0)) }
    }
}
